minidump-processor/src/op_analysis.rs

```diff
@@ -0,0 +1,733 @@
+/// a flat memory dump of process memory may implement this trait
+///
+pub trait SparseMemoryStream: Read + Seek {}
+
+/// Read the instruction bytes that were being run by the given thread
+///
+/// Use the given `context` to attempt to read `1 <= n <= MAX_INSTRUCTION_LENGTH`
+/// bytes at the instruction pointer from the byte stream passed in `memory`, which is any type
+/// that implements `Seek + Read` and is meant to represent the memory of the target machine
+///
+///
+///
+/// # Errors
+///
+/// This may fail if the underlying byte stream fails to `seek()` or `read()`
+///
+pub fn get_thread_instruction_bytes(
+    context: &MinidumpContext,
+    memory: &mut impl SparseMemoryStream,
+) -> Result<Vec<u8>, OpAnalysisError> {
+    let instruction_pointer = context.get_instruction_pointer();
+
+    memory
+        .seek(SeekFrom::Start(instruction_pointer))
+        .map_err(OpAnalysisError::ReadThreadInstructionFailed)?;
+
+    // We use MAX_INSTRUCTION_LENGTH here as an optimization to avoid allocating and copying
+    // more bytes than needed to produce an instruction
+
+    let mut buffer = Vec::with_capacity(MAX_INSTRUCTION_LENGTH.into());
+
+    memory
+        .take(MAX_INSTRUCTION_LENGTH.into())
+        .read_to_end(&mut buffer)
+        .map_err(OpAnalysisError::ReadThreadInstructionFailed)?;
+
+    Ok(buffer)
+}
+
+/// Details about a memory access performed by an instruction
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct MemoryAccess {
+    /// The address of the memory access
+    pub address: u64,
+    /// The size of the memory access
+    ///
+    /// Note that this is optional, as there are weird instructions that do not know the size
+    /// of their memory accesses without more complex context
+    pub size: Option<u8>,
+}
+
+/// Determine the memory accesses that would occur with the given instruction and context
+///
+/// Architectures like x86 allow for complex arithmetic involving multiple registers to be
+/// used to determine a target memory address, and allow for single operations involving
+/// multiple memory locations (bit blitting)
+///
+/// If the given `instruction_bytes` contain a valid instruction, and the given `context` contains
+/// valid values for all the registers used for address calculation, this function will return a
+/// (possibly-empty) list of all the memory accesses for the instruction
+///
+/// ⚠ NOTE ⚠ - Certain instructions like PUSH are not considered to access memory, even though
+/// they do technically write to the stack. Generally this shouldn't be of much concern, but
+/// there may be some situations where this detail matters
+///
+/// # Errors
+///
+/// May fail if the given bytes are too short to form an instruction, don't represent a valid
+/// instruction encoding, registers needed to calculate the address are invalid, or if the
+/// given CPU architecture is not supported or enabled by the current feature set
+///
+pub fn get_instruction_memory_access(
```

Comment:
This trailing blank doc-comment line is unconventional-enough style that I'd appreciate having them removed just for uniformity.

Reply:
Also I guess if I'm bothering you about comment formatting, ending paragraphs/sentences with periods would be ideal.

Reply:
Fair enough, and done!